sdk/python/kfp/components/pipeline_task.py

```diff
@@ -456,6 +470,56 @@ def my_pipeline():
             self._task_spec.dependent_tasks.append(task.name)
         return self
 
+    def set_exit_task(self, exit_task: 'PipelineTask') -> 'PipelineTask':
+        """Specifies the task to invoke after the current task exits,
+        irrespective of whether the current task completes successfully.
+
+        Args:
+            exit_task: Task to execute after this task is done
+
+        Example:
+          ::
+
+            @dsl.pipeline()
+            def my_pipeline(sample_input1: str = 'message'):
+                task = fail_op(message=sample_input1)
+                clean_up = print_op(message=task.output)
+                task.set_exit_task(exit_task=clean_up)
+        """
+        from kfp.components import pipeline_context
+        from kfp.components import tasks_group
+
+        for _, input_spec in exit_task.component_spec.inputs.items():
+            if not input_spec.optional:
+                raise ValueError(
+                    f'Clean up task {exit_task.name} requires a default value to make sure the Exit handler never fails.'
+                )
+
+        for task in pipeline_context.Pipeline.get_default_pipeline(
+        ).tasks.values():
+            if isinstance(
+                    task,
+                    tasks_group.ExitHandler) and task.exit_task == exit_task:
+                raise ValueError(
+                    f'Clean up task can not be used for multiple worker tasks. Clean up task {exit_task.name} already used for {task.name}'
+                )
+
+        pipeline_context.Pipeline.get_default_pipeline(
+        ).remove_task_from_groups(self)
+
+        with tasks_group.ExitHandler(exit_task=exit_task) as exit_handler:
+            pipeline_context.Pipeline.get_default_pipeline().add_task(
+                task=self,
+                add_to_group=not getattr(self, 'is_task_group_dependent',
+                                         False))
+
+        exit_handler._surface_outputs(self)
+        exit_task._change_inputs(exit_handler.outputs, exit_handler.name)
```

Comment:
I have modified the `_surface_outputs` code to be located in `compiler_utils.get_outputs_for_all_groups` however for `_change_inputs` those inputs need to be changed before `get_inputs_for_all_groups` should be called in `pipeline_spec` so that logic would need to be called before `get_outputs_for_all_groups` is executed, therefore it can not be moved from here, to combat the issue of avoiding leakage i added the actual logic to the compiler utils file

Reply:
I think it would be best to use the current [outputs defaultdict](https://github.com/kubeflow/pipelines/blob/eb5702ab391f9ffaff5b647364d0bb31effe1368/sdk/python/kfp/compiler/compiler_utils.py#L451) approach to handle surfacing outputs from an inner task to a parent `TasksGroup`. Specifically, creating more entries in this dictionary and removing special handling for this feature elsewhere in the codebase.

> however for _change_inputs those inputs need to be changed before get_inputs_for_all_groups

I see. I believe this would be addressed by aligning with the existing implementation in the  `get_outputs_for_all_groups` function. It's admittedly a detailed block of code to step through, but hopefully the comments help show how the current logic addresses this problem.

I'm happy to discuss/explain more as needed.